mlir/lib/Dialect/Rock/Transforms/Regularize.cpp

```diff
@@ -147,187 +146,452 @@ struct RegularizeGenericRewritePattern
+  if (!readers)
+    return;
+  SmallPtrSet<OpOperand *, 4> elementwiseArgs, onwardWrites;
+  for (OpOperand *reader : *readers) {
+    auto readOp = dyn_cast<MemoryEffectOpInterface>(reader->getOwner());
+    if (!readOp)
+      continue;
+    SmallVector<MemoryEffects::EffectInstance> effects;
+    readOp.getEffects(effects);
+    for (const auto &effect : effects) {
+      OpOperand *operand = effect.getEffectValue<OpOperand *>();
+      if (!operand || operand == reader)
+        continue;
+      if (isa<MemoryEffects::Write>(effect.getEffect())) {
+        onwardWrites.insert(operand);
+        elementwiseArgs.erase(operand);
       }
-      if (isFusorOp(useOp, result)) {
-        fusor = useOp;
-      } else if (auto lgop = dyn_cast<linalg::GenericOp>(useOp)) {
-        if (!fusor && llvm::is_contained(lgop.getOutputs(), result))
-          fusor = useOp;
+      if (isa<MemoryEffects::Read>(effect.getEffect()) &&
+          !onwardWrites.contains(operand)) {
+        elementwiseArgs.insert(operand);
       }
     }
+  }
+  for (OpOperand *arg : elementwiseArgs)
+    collectInputFusionWrites(arg, bufferDeps, state);
+  for (OpOperand *localWrite : onwardWrites) {
+    LLVM_DEBUG(llvm::dbgs() << "Traversing via onward writer: "
+                            << *localWrite->getOwner() << "\n");
+    collectOutputFusionWrites(localWrite, bufferDeps, state);
+  }
+}
 
-    // find fusee's transform chains
-    for (auto &use : buffer.getUses()) {
-      Operation *useOp = use.getOwner();
-      SmallVector<Operation *> chain;
-      bool isWriter = collectChain(buffer, useOp, chain);
-      if (isWriter) {
-        assert(writer == nullptr);
-        writer = useOp;
-      }
-      if (!chain.empty() && chain.back() != fusor) {
-        hasTransforms |= chain.size() > 1;
-        readChains.push_back(chain);
-      }
+////////////////////////////////////////////////////////////////////////
+////  Push Transforms Over alloc to fusor
+////////////////////////////////////////////////////////////////////////
+
+/// In the below, a fusor is the operation that will have some computation fused
+/// into it. This is usually a gemm (either its output or inputs) but can be,
+/// for example, another linalg.generic. The operation that will be fused into
+/// the fusor is the fusee.
+///
+/// A fusor writes to or reads from a memref.alloc() and a fusee does the
+/// opposite of what the fusor does. Before this pass runs, both the fusor and
+/// the fusee may access that intermediate buffer (which will be rewritten away
+/// by fusion) through some sequence of view operations.
+///
+/// After these rewrites, we have the following invariants, which make the
+/// fusion code itself (it's down in AlignTiling.cpp) much simpler and enables
+/// analysis of vectorization within the parts of the pipeline between this pass
+/// and AlignTiling (which can be important for key performance queries).
+///
+/// - The fusee always has a direct reference to the buffer
```

Comment:
please define buffer. 

This should be the common buffer between fusee and fusor.